tlsf_cpp/include/tlsf_cpp/tlsf.hpp

```diff
@@ -0,0 +1,148 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// C++ wrapper for Miguel Masmano Tello's implementation of the TLSF memory allocator
+// Implements the allocator_traits template
+
+#ifndef TLSF_CPP__TLSF_HPP_
+#define TLSF_CPP__TLSF_HPP_
+
+#include <cstring>
+#include <iostream>
+#include <new>
+#include <stdexcept>
+
+#include "tlsf/tlsf.h"
+
+// Custom allocator deleter
+
+// Necessary for using custom allocator with std::basic_string in GCC 4.8
+// See: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=56437
+template<typename T>
+struct allocator_pointer_traits
+{
+  using size_type = std::size_t;
+  using pointer = T *;
+  using const_pointer = const pointer;
+  using difference_type = typename std::pointer_traits<pointer>::difference_type;
+  using reference = T &;
+  using const_reference = const reference;
+};
+
+template<>
+struct allocator_pointer_traits<void>
+{
+};
+
+template<typename T>
+struct tlsf_heap_allocator : public allocator_pointer_traits<T>
+{
+  // Needed for std::allocator_traits
+  using value_type = T;
+
+  explicit tlsf_heap_allocator(size_t size)
+  : memory_pool(nullptr), pool_size(size)
+  {
+    initialize(size);
+  }
+
+  // Needed for std::allocator_traits
+  tlsf_heap_allocator()
+  : memory_pool(nullptr)
+  {
+    initialize(1024 * 1024);
```

Comment:
Well...

One annoying thing about how the C++ standard library handles allocators is that many data structures that take an allocator, such as `std::vector`, take the allocator type as a template argument, and then default-construct their allocator in their own constructor.

So most instantiations of the allocator will use this default constructor, which is really not awesome.

I previously had the idea to add a template argument to the allocator so that the user has a better path to initializing an allocator, even one that gets default-constructed. I can try implementing that now...